File: src/types/question.ts
export interface Question {
  id: string
  questionType: QuestionType | null
  stem: string
  correctAnswer: string
  explanation: string | null
  note: string | null
  createdAt: string
  updatedAt: string
  sourceId: string | null
  tagIds: string[]
  attachmentIds: string[]
}

export type QuestionType =
  | 'SINGLE_SELECT'
  | 'MULTIPLE_SELECT'
  | 'TRUE_FALSE'
  | 'FILL_IN_THE_BLANK'
  | 'SHORT_ANSWER'
  | 'CALCULATION'
  | 'ESSAY'

export interface CreateQuestionRequest {
  sourceId?: string | null
  questionType?: QuestionType | null
  stem: string
  correctAnswer: string
  explanation?: string | null
  note?: string | null
  tagIds?: string[]
  attachmentIds?: string[]
}

export interface CreateQuestionResponse {
  question: Question
}

export interface UpdateQuestionRequest {
  id: string
  sourceId: string | null
  questionType: QuestionType | null
  stem: string
  correctAnswer: string
  explanation: string | null
  note: string | null
  tagIds: string[]
  attachmentIds: string[]
}

export interface UpdateQuestionResponse {
  question: Question
}

export interface DeleteQuestionRequest {
  id: string
}

export interface DeleteQuestionResponse {
  id: string
}

export interface GetQuestionRequest {
  id: string
}

export interface GetQuestionResponse {
  question: Question
}

export interface ListQuestionsRequest {
  filter?: QuestionFilter
  sort?: QuestionSort[]
  offset?: number
  limit?: number
}

export type QuestionSort =
  | 'UPDATED_AT_ASC'
  | 'UPDATED_AT_DESC'
  | 'MASTERY_ASC'
  | 'MASTERY_DESC'
  | 'ID_ASC'
  | 'ID_DESC'

export interface QuestionFilter {
  search?: string
  book?: string
  chapter?: string
  knowledge?: string
  tagIds?: string[]
  updatedSince?: string
  reviewState?: 'DUE' | 'NOT_DUE'
}

export interface ListQuestionsResponse {
  items: Question[]
  total: number
}

File: src-tauri/src/command/request/question.rs
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CreateQuestionRequest {
    pub(crate) source_id: Option<String>,
    pub(crate) question_type: Option<String>,
    pub(crate) stem: String,
    pub(crate) correct_answer: String,
    pub(crate) explanation: Option<String>,
    pub(crate) note: Option<String>,
    #[serde(default)]
    pub(crate) tag_ids: Vec<String>,
    #[serde(default)]
    pub(crate) attachment_ids: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UpdateQuestionRequest {
    pub(crate) id: String,
    #[serde(deserialize_with = "super::deserialize_required_nullable")]
    pub(crate) source_id: Option<String>,
    #[serde(deserialize_with = "super::deserialize_required_nullable")]
    pub(crate) question_type: Option<String>,
    pub(crate) stem: String,
    pub(crate) correct_answer: String,
    #[serde(deserialize_with = "super::deserialize_required_nullable")]
    pub(crate) explanation: Option<String>,
    #[serde(deserialize_with = "super::deserialize_required_nullable")]
    pub(crate) note: Option<String>,
    pub(crate) tag_ids: Vec<String>,
    pub(crate) attachment_ids: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeleteQuestionRequest {
    pub(crate) id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct GetQuestionRequest {
    pub(crate) id: String,
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ListQuestionsRequest {
    #[serde(default)]
    pub(crate) filter: QuestionFilterRequest,
    pub(crate) sort: Option<Vec<String>>,
    pub(crate) offset: Option<usize>,
    pub(crate) limit: Option<usize>,
}

#[derive(Debug, Default, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct QuestionFilterRequest {
    pub(crate) search: Option<String>,
    pub(crate) book: Option<String>,
    pub(crate) chapter: Option<String>,
    pub(crate) knowledge: Option<String>,
    pub(crate) tag_ids: Option<Vec<String>>,
    pub(crate) updated_since: Option<String>,
    pub(crate) review_state: Option<String>,
}
